gpu4pyscf/scf/jk.py

```diff
@@ -0,0 +1,669 @@
+    _env = mol._env.copy()
+    ls = mol._bas[:,ANG_OF]
+    ptr, idx = np.unique(mol._bas[:,PTR_COEFF], return_index=True)
+    ptr = ptr[ls[idx] < 2]
+    idx = idx[ls[idx] < 2]
+    fac = ((ls[idx]*2+1) / (4*np.pi)) ** .5
+    nprim = mol._bas[idx,NPRIM_OF]
+    nctr = mol._bas[idx,NCTR_OF]
+    for p, n, f in zip(ptr, nprim*nctr, fac):
+        _env[p:p+n] *= f
+    return _env
+
+def iter_cart_xyz(n):
+    return [(x, y, n-x-y)
+            for x in reversed(range(n+1))
+            for y in reversed(range(n+1-x))]
+
+def g_pair_idx(ij_inc=None):
+    dat = []
+    xyz = [np.array(iter_cart_xyz(li)) for li in range(LMAX+1)]
+    for li in range(LMAX+1):
+        for lj in range(LMAX+1):
+            li1 = li + 1
+            idx = (xyz[lj][:,None] * li1 + xyz[li]).transpose(2,0,1)
+            dat.append(idx.ravel())
+    g_idx = np.hstack(dat).astype(np.int32)
+    offsets = np.cumsum([0] + [x.size for x in dat]).astype(np.int32)
+    return g_idx, offsets
+
+def init_constant(mol):
+    g_idx, offsets = g_pair_idx()
+    libvhf_rys.RYS_init_constant(
+        g_idx.ctypes, offsets.ctypes, mol._env.ctypes, ctypes.c_int(mol._env.size),
+        ctypes.c_int(SHM_SIZE))
+
+def _make_j_engine_pair_locs(mol):
+    ls = mol._bas[:,ANG_OF]
+    ll = (ls[:,None]+ls).ravel()
+    pair_loc = np.append(0, np.cumsum((ll+1)*(ll+2)*(ll+3)//6))
+    return np.asarray(pair_loc, dtype=np.int32)
+
+def quartets_scheme(mol, l_ctr_pattern, shm_size=SHM_SIZE):
+    ls = l_ctr_pattern[:,0]
+    li, lj, lk, ll = ls
+    order = li + lj + lk + ll
+    nfi = (li + 1) * (li + 2) // 2
+    nfj = (lj + 1) * (lj + 2) // 2
+    nfk = (lk + 1) * (lk + 2) // 2
+    nfl = (ll + 1) * (ll + 2) // 2
+    gout_size = nfi * nfj * nfk * nfl
+    if (gout_size <= UNROLL_NFMAX or order <= UNROLL_ORDER) and all(ls <= UNROLL_LMAX):
+        if (CUDA_VERSION >= 12040 and
+            order <= 3 and (li,lj,lk,ll) != (1,1,1,0) and (li,lj,lk,ll) != (1,0,1,1)):
+            return 512, 1
+        return 256, 1
+
+    g_size = (li+1)*(lj+1)*(lk+1)*(ll+1)
+    nps = l_ctr_pattern[:,1]
+    ij_prims = nps[0] * nps[1]
+    nroots = (li+lj+lk+ll) // 2 + 1
+
+    if mol.omega >= 0:
+        unit = nroots*2 + g_size*3 + ij_prims*4
+    else: # SR
+        unit = nroots*4 + g_size*3 + ij_prims*4
+    counts = shm_size // (unit*8)
+    n = 128
+    while n >= counts:
+        n >>= 1
+    gout_stride = THREADS // n
+    while gout_stride < 128 and gout_size / (gout_stride*GOUT_WIDTH) > 1:
+        n //= 2
+        gout_stride *= 2
+    return n, gout_stride
+
+def _j_engine_quartets_scheme(mol, l_ctr_pattern, shm_size=SHM_SIZE):
+    ls = l_ctr_pattern[:,0]
+    li, lj, lk, ll = ls
+    order = li + lj + lk + ll
+    lij = li + lj
+    lkl = lk + ll
+    nmax = max(lij, lkl)
+    nps = l_ctr_pattern[:,1]
+    ij_prims = nps[0] * nps[1]
+    g_size = (lij+1)*(lkl+1)
+    nf3_ij = (lij+1)*(lij+2)*(lij+3)/6
+    nf3_kl = (lkl+1)*(lkl+2)*(lkl+3)/6
+    nroots = (li+lj+lk+ll) // 2 + 1
+    lmax = 4
```

Comment:
This is hardcoded?

Reply:
It's not the angular moment of orbital basis. This is a special/temporary treatment for j-engine.